server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -40,26 +41,64 @@
 
     @Override
     public byte readByte() throws IOException {
-        if (offset() >= bytesReference.length()) {
```

Comment:
no need for double checking here, the slice lengths must be correct (it's an assumption we make in other places so the `maybeNextSlice` is all we need